pkgs/stdenv/generic/jscall.c

```diff
@@ -0,0 +1,224 @@
+            perror("jscall token return failed");
+        }
+        break;
+    }
+}
+
+static void handle_signal(int sig)
+{
+    if (w >= 0 && token != 0) {
+        return_token();
+        w = -1;
+    }
+    signal(sig, SIG_DFL);
+    raise(sig);
+}
+
+static int find_separator(char** argv)
+{
+    for (int i = 0; argv[i]; i++) {
+        if (strcmp(argv[i], "----") == 0) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+static void shift_args(char** low, char** high)
+{
+    for (; high[0]; low++, high++) {
+        low[0] = high[0];
+    }
+    low[0] = NULL;
+}
+
+static void shift_once(char** args)
+{
+    shift_args(args, args + 1);
+}
+
+static int run_child(char** argv)
+{
+    execvp(argv[0], argv);
+    perror("execvp");
+    return 1;
+}
+
+static const char* usage =
+    "invocation: jscall [-F flags] <jobserver> <token-limit> <main...> ---- [def-args...] ---- [args...]\n"
+    "\n"
+    "runs `main... args...` with a GNU jobserver fd set in MAKEFLAGS if `jobserver`\n"
+    "exists on is a nixos-jobserver endpoint. otherwise runs\n"
+    "`main... def-args... args...`. if -F is given, `flags` is a space-separated\n"
+    "of environment variables to add jobserver arguments to instead of MAKEFLAGS.\n"
+    "\n"
+    "when running with nixos-jobserver available `main` and its children are\n"
+    "limited to `token-limit` tokens, which may be less than the system limit.\n"
+    "\n"
+    "arguments beginning with ---- are not allowed except in `args`.\n";
+
+int main(int argc, char* argv[])
+{
+    // we'll later pass this to strtok, which needs it to be mutable
+    char vars_raw[] = "MAKEFLAGS";
+    char* vars = vars_raw;
+
+    for (int opt; (opt = getopt(argc, argv, "+F:")) != -1; ) {
+        switch (opt) {
+        case 'F':
+            vars = optarg;
+            break;
+        case '?':
+        default:
+            fprintf(stderr, "%s", usage);
+            return 1;
+        }
+    }
+
+    shift_args(argv + 1, argv + optind);
+    argc -= optind - 1;
+
+    // create a default command first by removing the separators.
+    // if we can use the jobserver we'll shift the default args out of the arguments
+    // list.
+
+    char** cmd = argv + 3;
+    int cmd_length = find_separator(cmd);
+    if (cmd_length < 0) {
+        fprintf(stderr, "main command not terminated\n\n%s", usage);
+        return 1;
+    } else if (cmd_length == 0) {
+        fprintf(stderr, "main command not given\n\n%s", usage);
+        return 1;
+    }
+    shift_once(cmd + cmd_length);
+
+    char** def_args = cmd + cmd_length;
+    int def_args_length = find_separator(def_args);
+    if (def_args_length < 0) {
+        fprintf(stderr, "def-args command not terminated\n\n%s", usage);
+        return 1;
+    }
+    shift_once(def_args + def_args_length);
+
+    char** args = def_args + def_args_length;
+
+    struct stat st;
+    if (stat(argv[1], &st)) {
+        perror("stat() jobserver pipe");
```

Comment:
technically this could race if the socket vanishes in the meantime; I'm not awake enough at the moment to be certain of whether the umount would cause that symptom.